l10n_it_intrastat_statement/models/intrastat.py

```diff
@@ -0,0 +1,2163 @@
+        'account.intrastat.statement.purchase.section3',
+        'statement_id', string='Purchase - Section 3')
+    purchase_section3_operation_number = fields.Integer(
+        store=True, readonly=True,
+        compute='_compute_amount_purchase_s3')
+    purchase_section3_operation_amount = fields.Integer(
+        store=True, readonly=True,
+        compute='_compute_amount_purchase_s3')
+    purchase_section4_ids = fields.One2many(
+        'account.intrastat.statement.purchase.section4',
+        'statement_id', string='Purchase - Section 4')
+    purchase_section4_operation_number = fields.Integer(
+        store=True, readonly=True,
+        compute='_compute_amount_purchase_s4')
+    purchase_section4_operation_amount = fields.Integer(
+        store=True, readonly=True,
+        compute='_compute_amount_purchase_s4')
+
+    @api.model
+    def create(self, vals):
+        statement = super().create(vals)
+        statement._normalize_statement()
+        return statement
+
+    @api.multi
+    def write(self, vals):
+        res = super().write(vals)
+        self._normalize_statement()
+        self.recompute_sequence_lines()
+        return res
+
+    @api.depends('fiscalyear', 'period_type', 'period_number')
+    @api.onchange('fiscalyear', 'period_type', 'period_number')
+    def _compute_dates(self):
+        year = self.fiscalyear
+        if not year \
+           or not self.period_type \
+           or not self.period_number:
+            return
+
+        period_error_msg = self.check_period_number()
+        if period_error_msg:
+            return {'warning': {'message': period_error_msg}}
+
+        if self.period_type == 'M':
+            month = self.period_number
+            period_date_start = datetime(year, month, 1)
+            period_date_stop = \
+                datetime(year, month, 1) \
+                + relativedelta(months=1) \
+                - timedelta(days=1)
+        else:
+            quarter = self.period_number
+            month_start = int(12 / 4 * (quarter - 1) + 1)
+            period_date_start = datetime(year, month_start, 1)
+            period_date_stop = \
+                period_date_start \
+                + relativedelta(months=3) \
+                - timedelta(days=1)
+        self.date_start = fields.Date.to_date(period_date_start)
+        self.date_stop = fields.Date.to_date(period_date_stop)
+
+    @api.model
+    def _get_period_ref(self, invoice_line):
+        res = {
+            'year_id': False,
+            'quarterly': False,
+            'month': False
+        }
+
+        res.update({'year_id': self.fiscalyear})
+        # Month/Quarter
+        if self.period_type == 'T':
+            res.update({'quarterly': self.period_number})
+        else:
+            res.update({'month': self.period_number})
+        return res
+
+    def _normalize_statement(self):
+        # Unlink lines sale/purchase sections
+        self.ensure_one()
```

Comment:
Viene chiamato da `write`, quindi necessita di un ciclo `for` su `self` e, per chiarezza, aggiungerei il decoratore `@api.multi`